hack/generator/pkg/codegen/swagger_type_extractor.go

```diff
@@ -0,0 +1,170 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package codegen
+
+import (
+	"context"
+	"regexp"
+	"strings"
+
+	"github.com/Azure/k8s-infra/hack/generator/pkg/astmodel"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/config"
+	"github.com/Azure/k8s-infra/hack/generator/pkg/jsonast"
+	"github.com/go-openapi/spec"
+	"github.com/pkg/errors"
+	"k8s.io/klog/v2"
+)
+
+type typeExtractor struct {
+	idFactory astmodel.IdentifierFactory
+	config    *config.Configuration
+	cache     *jsonast.OpenAPISchemaCache
+	// group for output types (e.g. Microsoft.Network.Frontdoor)
+	outputGroup   string
+	outputVersion string
+}
+
+// extractTypes finds all operations in the Swagger spec that
+// have a PUT verb and a path like "Microsoft.GroupName/…/resourceName/{resourceId}",
+// and extracts the types for those operations, into the 'resources' parameter.
+// Any additional types required by the resource types are placed into the 'otherTypes' parameter.
+func (extractor *typeExtractor) extractTypes(
+	ctx context.Context,
+	filePath string,
+	swagger spec.Swagger,
+	resources astmodel.Types,
+	otherTypes astmodel.Types) error {
+
+	packageName := extractor.idFactory.CreatePackageNameFromVersion(extractor.outputVersion)
+
+	scanner := jsonast.NewSchemaScanner(extractor.idFactory, extractor.config)
+
+	for operationPath, op := range swagger.Paths.Paths {
+		put := op.Put
+		if put == nil {
+			continue
+		}
+
+		resourceName, err := extractor.resourceNameFromOperationPath(packageName, operationPath)
+		if err != nil {
+			klog.Error(err)
+			continue
+		}
+
+		resourceType, err := extractor.resourceTypeFromOperation(ctx, scanner, swagger, filePath, put)
+		if err != nil {
+			if err == context.Canceled {
+				return err
+			}
+
+			klog.Errorf("unable to produce type for %v: %v", resourceName, err)
+			continue // TODO: make fatal
```

Comment:
Is this TODO for later or for before you merge?

Reply:
later; need to address some items in #219

Reply:
I'll add a specific item about this.